Redis Vector Store

Spartee opened the pull request:
### Description

PR adds Redis + RediSearch as a supported vector database. This is a new PR which replaces #1241 as that PR was based on a prior version of Llama_index which has since been refactored.

This should also close #452 

### TODO (carried over)

- [x] docs
- [x] Fix document relationship
- [x] Return similarities
- [x] Better connection handling
- [x] fix importing 
- [x] Code clean up and comments
- [x] Example

Spartee commented:
FYI @jerryjliu @Disiok I went ahead a did a new version. Overall the interface looks much better y'all. little more to do but almost done. Would love input.

Disiok commented on gpt_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,313 @@
+"""Redis Vector store index.
+
+An index that that is built on top of an existing vector store.
+"""
```
nit: this should just be vector store now, no need to mention index anymore after refactor.

Disiok commented on gpt_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,313 @@
+
+import numpy as np
+
+from gpt_index.data_structs.node_v2 import DocumentRelationship, Node
```
we finally committed to replacing all `gpt_index` with `llama_index`. 

Sorry for the myriad of refactoring changes!

Disiok commented on gpt_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,313 @@
+        query_params = {"vector": np.array(query.query_embedding).astype(np.float32).tobytes()}
+        _logger.info(f"Querying index {self._index_name}")
+
+        try:
+            results = self._redis_client.ft(self._index_name).search(redis_query,
```
nit: missing except?

Disiok commented on gpt_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,313 @@
+        for doc in results.docs:
+            node = Node(
+                text=doc.text,
+                doc_id=doc.doc_id.replace("_", "-"),
```
this is super confusing and our fault, but `Node.doc_id` is the ID of the node, `Node.ref_doc_id` is the source document ID (set via `{DocumentRelationship.SOURCE: doc.doc_id}`).

So here we should restore with `doc_id=doc.id` I believe.

Spartee commented on gpt_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,313 @@
+        return self._index_name in indices
+
+    def _prep_query(self, top_k: int = 10) -> str:
+        # returns vector_score by default
```
fixed

Spartee commented:
Thanks for the review @softboyjimbo and @Disiok. I've fixed the similarities, used the correct escape for querying on delete and cleaned up the init a little bit. Will address other refactor related fixes and docs soon.

softboyjimbo commented on gpt_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,310 @@
+            )
+            ids.append(doc.id)
+            nodes.append(node)
+            scores.append(1 - float(doc.vector_score))
```
lgtm

Disiok commented:
Hey @Spartee  just checking, is this ready for another review? Happy to help get this in.

Spartee commented:
@Disiok ready now! I believe I hit each point of the review, but let me know!

Spartee commented:
Whoops, API key deleted. Thanks GitGuardian

Disiok commented on .vscode/settings.json:
```diff
@@ -1,7 +1,10 @@
-    },
-}
\ No newline at end of file
+  "python.formatting.provider": "none",
+  "editor.formatOnSave": true,
+  "editor.codeActionsOnSave": {
+    "source.organizeImports": true
+  },
+  "[python]": {
+    "editor.defaultFormatter": "ms-python.black-formatter"
+  }
+}
```
Could you remove this. Guessing it just leaked in.

Disiok commented on llama_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,323 @@
+
+        return VectorStoreQueryResult(nodes=nodes, ids=ids, similarities=scores)
+
+    def persist(self, in_background=True) -> None:
```
right now `VectorStore` defines `persist` with a `persist_path` param:
```python 
    def persist(self, persist_path: str) -> None:
        return None
```
This is because we provide a convenience method to persist to disk via `StorageContext`, which tries to save everything into the same folder by configuring and passing the persist_path.

Could we just take in the arg and ignore it here?

Spartee replied:
done

Spartee replied:
wasn't exactly sure what you wanted/mypy wanted here, but I took a crack at it. Happy to make this a different method if that's what you meant.

Spartee replied:
wasn't exactly sure what you wanted/mypy wanted here, but I took a crack at it. Happy to make this a different method if that's what you meant.

Disiok commented on llama_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,323 @@
+    def _create_vector_field(
+        self,
+        name: str,
+        dims: int = "1536",
```
nit: passing str to int?

Spartee replied:
done

Disiok commented on llama_index/vector_stores/redis.py:
```diff
@@ -0,0 +1,323 @@
+        )
+        return redis_query
+
+    def _create_vector_field(
```
It wasn't immediately obvious to me that users can pass in index args to configure this. Is this documented somewhere else? Do you expect the users to change these?

Spartee replied:
In most cases, no. But advanced users may want to, hence the approach. Added pointers to the docs. They are named the exact same so it's easy to recognize for people who know what they are doing. For others, it'll be a footgun.

Disiok replied:
great

Spartee commented:
@Disiok looks like this should be good to go (at least workflow approval)

Spartee commented:
@Disiok this looks like the linting errors are all due to not having the redis-types library installed.

Spartee commented:
@jerryjliu @Disiok, thoughts here?

jerryjliu commented:
> @jerryjliu @Disiok, thoughts here?

Thanks for raising @Spartee. I think lazy type imports should help fix this (e.g. type client as Any in function signature and then cast to type within function). Do you have alternative suggestions given the current import approach?

Spartee commented:
@jerryjliu That feels like "why am I using mypy then?" running ``mypy --install-types .`` instead of ``mypy .`` would address it as it stands I believe. 

I used the ``If TYPE_CHECKING`` guard, but it didn't seem to solve that particular issue.

I'd rather use the actual types if possible, but really I just want to get this in as I have a project waiting for it so either way is fine with me.

Disiok commented:
> @jerryjliu That feels like "why am I using mypy then?" running `mypy --install-types .` instead of `mypy .` would address it as it stands I believe.
> 
> I used the `If TYPE_CHECKING` guard, but it didn't seem to solve that particular issue.
> 
> I'd rather use the actual types if possible, but really I just want to get this in as I have a project waiting for it so either way is fine with me.

@Spartee We'll get it in tonight for sure!
